### PR TITLE
[WT-1012][DO NOT MERGE] WalletConnection Changes Spike

### PR DESCRIPTION
**Pull request by MitchSaadIMX**
…nection

# Summary

# Why the changes

# Things worth calling out

# Before merging
- [ ] ***For Immutable developers:*** Do the changes in this PR require documentation updates? Please refer to this [SDK documentation guide](https://immutable.atlassian.net/wiki/spaces/PPS/pages/1916994017/SDK+documentation+guide) and list links to documentation update PRs (they don't have to be merged) below (or write `N/A`):
    - [ ] *Add documentation update 1*
    - [ ] *Add documentation update 2*

### CODE REVIEW

File: examples/completeEthWithdrawal.ts [hunk 1]
```diff
@@ -7,10 +7,10 @@ import { generateWalletConnection } from './libs/walletConnection';
 
     const client = new ImmutableX(Config.SANDBOX);
 
-    const starkPublicKey = await walletConnection.starkSigner.getAddress();
+    const starkPublicKey = await walletConnection.signers.starkExSigner.getAddress();
 
     const completeWithdrawalResponse = await client.completeWithdrawal(
-      walletConnection.ethSigner,
+      walletConnection.signers.ethSigner,
```
Comment (cajames) [inline examples/completeEthWithdrawal.ts:13]: My understanding is that this is what could have been done with all the existing workflows

**Review by cajames — COMMENTED** [review]
Hey Mitch, thanks for sending this through. Can we please meet and discuss this a bit more? I'd like to understand better the changes made to the "signRaw" function. Especially why we're signing a signature?

And also, let's discuss these changes to the function params.

File: src/utils/crypto/crypto.ts [hunk 2]
```diff
@@ -41,7 +42,9 @@ export async function signRaw(
   payload: string,
   signer: Signer,
 ): Promise<string> {
-  const signature = deserializeSignature(await signer.signMessage(payload));
+  const signature = deserializeSignature(
+    await signer.signMessage({ payload: '', message: payload, signature: '' }),
```
[the remark this answers is not in the document]
Comment (MitchSaadIMX) [reply src/utils/crypto/crypto.ts:46]: payload here is the message hash - i should have done a bit of renaming here to make it clearer. 

the payload attribute also shouldn't be a string, thats my bad - ill clean this up a bit, but the idea is that the signer gets 

- `payload` the json object representing the transaction
- `message` is the message hash
- `signature` is how they verify that the transaction payload they are showing to the user matches the message hash

what we would get back is the signature like normal

File: src/workflows/withdrawal/prepareWithdrawal.ts [hunk 3]
```diff
@@ -29,7 +40,11 @@ export async function prepareWithdrawalWorkflow(
   const { signable_message: signableMessage, payload_hash: payloadHash } =
     signableWithdrawalResult.data;
 
-  const starkSignature = await starkSigner.signMessage(payloadHash);
+  const starkSignature = await starkExSigner.signMessage({
+    payload: '',
+    message: payloadHash,
+    signature: '',
+  });
```
[the remark this answers is not in the document]
Comment (MitchSaadIMX) [reply src/workflows/withdrawal/prepareWithdrawal.ts:47]: they are just placeholders - see above

File: src/workflows/withdrawal/prepareWithdrawal.ts [hunk 4]
```diff
@@ -8,15 +9,25 @@ const assertIsDefined = <T>(value?: T): T => {
...
+};
 
 export async function prepareWithdrawalWorkflow(
   params: PrepareWithdrawalWorkflowParams,
 ): Promise<CreateWithdrawalResponse> {
-  const { ethSigner, starkSigner, withdrawalsApi } = params;
+  const {
+    walletConnection: {
+      signers: { ethSigner, starkExSigner },
+    },
```
[the remark this answers is not in the document]
Comment (MitchSaadIMX) [reply src/workflows/withdrawal/prepareWithdrawal.ts:23]: The idea is that the consumer never has to worry about whats in the WalletConnection - they just pass it through and we take the signers we need from it. At the moment its just the eth and starkex signers every time 😅

[the remark this answers is not in the document]
Comment (MitchSaadIMX) [reply src/workflows/withdrawal/prepareWithdrawal.ts:23]: also - even if we just passed the signers, the signMessage function signatures have changes

File: src/utils/crypto/crypto.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (suluvai) [reply src/utils/crypto/crypto.ts:46]: I am completely lost why we have verification in `signMessage` function?

File: src/workflows/withdrawal/prepareWithdrawal.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (suluvai) [reply src/workflows/withdrawal/prepareWithdrawal.ts:47]: Not convinced why those additional params are required?

**Pull request closed without merging**